### ISSUE TITLE
Structured headers: serializing token vs string

### ISSUE THREAD

**Issue by evert** [httpwg/http-extensions#790]
Hi everyone!

Draft-10 contains the following excerpt:

>    3.  If input_item is a string, return the result of applying
>        Serialising a String (Section 4.1.8) to input_item.
> 
>    4.  If input_item is a token, return the result of Serialising a
>        Token (Section 4.1.9) to input_item.

I am currently writing a [javascript implementation](https://github.com/evert/structured-headers), where there is no real native 'token' type (I suspect that this is true for most languages).

I'm curious, as an implementer, what should I do with that sentence? I have a few options:

1. Treat any string that can be represented as a token by serializing it as a token, but anything else as a string.
2. Treat everything as a string and ignore the token
3. Expose a sort of wrapper that marks certain strings as tokens.

Even though I feel that 1 is the nicest user experience, I worry that this will cause issues for future use-cases of structured headers that explicitly demand certain members to be serialized as either a token or string.

Can you give a recommendation on what I should do?

**Comment by phluid61** [comment]
We discussed this in #505; I think it's cycled back around to where your generic API needs a piece of metadata alongside stringy data to say whether to use token or string.

Whether that metadata is in the form of a second parameter, or a wrapper class, or some other signal... /shrug

In my [naive pure-Ruby implementation](https://github.com/phluid61/http-structured-headers) I created wrapper classes, so either a `Symbol` or an instance of my custom `Identifier` (needs to be renamed) class will be encoded as a token.  I don't know how well that works in the real world.

**Comment by mnot** [comment]
What @phluid61 said. Perhaps it's worth adding some implementation considerations along these lines?

**Comment by evert** [comment]
One of the things I have a bit of a hard time with here... is that I find it difficult to make a real case for why anyone would want the flexibility to choose either a string or token in cases where either may be used.

I can understand in some cases wanting to be able to restrict the character set to a few safe characters, and I can understand that in some cases you would want everything that 'string' affords, but the second case, I'm not sure if I see the benefit of users to have to make an explicit choice.

In other words, how open would you be to altering the specification a bit to allow any string to be encoded as a token if the character set allows it?

That way we can basically treat strings and tokens as identical when either may appear, and force tokens where only tokens are allowed.

**Comment by phluid61** [comment]
> One of the things I have a bit of a hard time with here... is that I find it difficult to make a real case for why anyone would want the flexibility to choose either a string or token in cases where either may be used.

Strings and tokens are fundamentally different things, the similarity of their serialisation notwithstanding. A string is a string; character data; unconstrained.  A token is a symbolic identifier; a member of a set; atomic.  So it shouldn't ever be the case that "either may be used."

When someone is writing a spec and has to choose between the two: If communicating, say, a filename, then use a string.  If communicating an identifier that represents which encoding to use, then use a token.

> That way we can basically treat strings and tokens as identical when either may appear, and force tokens where only tokens are allowed.

I don't understand.  Either tokens are just strings that don't need quotes to be serialised, in which case there'd never be a time when they are allowed but other strings/binaries aren't; or they are different things, so really oughtn't to be de/serialised ambiguously.

**Comment by evert** [comment]
You are describing the motivation behind the current state of the spec. My question is, can this be changed?

**Comment by reschke** [comment]
bq. I don't understand. Either tokens are just strings that don't need quotes to be serialised, in which case there'd never be a time when they are allowed but other strings/binaries aren't;  (...)

In current header fields, this is usally the case (everywhere we currently say "token / quoted-string" right now). Why change that?

**Comment by phluid61** [comment]
The only places I know if where we do that, it's in a "x=y" parameter or extension point, where the _y_ could be either a string or a token. *Not* because there's no difference between a string and a token.

We achieve that in SH by replacing `token / quoted-string` with `sh-item`

**Comment by mnot** [comment]
While there are cases like parameters where tokens and quoted strings are equivalent, there are other places where it matters; we re-introduced tokens (nee identifiers) to allow for some WHATWG-defined headers to be potentially ported to SH, and they have situations where a token needs to be a token (i.e., unquoted).

**Comment by reschke** [comment]
FWIW: I think in general we should allow token where quote-string is allowed (as simplified syntax), but not the other way around.

**Comment by evert** [comment]
Yeah I agree with @reschke, so @mnot's use case should still be fine.

**Comment by phluid61** [comment]
If string and token are, as I believe, different _types_, then this part RFC 7231 is a legacy Bad Thing:tm:, and since Structured Headers is a clean break of sorts, we are under no obligation to persist it:

```
   A parameter value that matches the token production can be
   transmitted either as a token or within a quoted-string.  The quoted
   and unquoted values are equivalent.  For example, the following
   examples are all equivalent, but the first is preferred for
   consistency:

     text/html;charset=utf-8
     text/html;charset=UTF-8
     Text/HTML;Charset="utf-8"
     text/html; charset="utf-8"
```

If string and token really are the same thing, then it's hard to justify any syntactic location that allows `sh-token` but not `sh-string`.  In terms of implementation, you'd still have to feed that "encode as token" metadata into the serialiser, or force every serialiser to never put quotes around strings that contain only `sh-token`-compatible characters.

\<glib>
Extrapolating _ad absurdum_ we might as well remove token completely, from everywhere. `content-type: "text/html"; "charset"="utf-8"`

... I mean, it works for JSON.
\</glib>

**Comment by mnot** [comment]
I'm inclined to close this with no action.
